types.go

```diff
@@ -116,21 +118,68 @@ type FunctionInfo struct {
 type functionInfo struct {
 	FunctionInfo
 
-	function         Function
-	gradient         Gradient
-	functionGradient FunctionGradient
-	statuser         Statuser
+	function                Function
+	gradient                Gradient
+	hessian                 Hessian
+	functionGradient        FunctionGradient
+	functionGradientHessian FunctionGradientHessian
+	statuser                Statuser
+}
+
+func newFunctionInfo(f Function) *functionInfo {
+	gradient, isGradient := f.(Gradient)
+	hessian, isHessian := f.(Hessian)
+	funcGrad, isFuncGrad := f.(FunctionGradient)
+	funcGradHess, isFuncGradHess := f.(FunctionGradientHessian)
+	statuser, isStatuser := f.(Statuser)
+
+	return &functionInfo{
+		FunctionInfo: FunctionInfo{
+			IsGradient:                isGradient,
+			IsHessian:                 isHessian,
+			IsFunctionGradient:        isFuncGrad,
+			IsFunctionGradientHessian: isFuncGradHess,
+			IsStatuser:                isStatuser,
+		},
+		function:                f,
+		gradient:                gradient,
+		hessian:                 hessian,
+		functionGradient:        funcGrad,
+		functionGradientHessian: funcGradHess,
+		statuser:                statuser,
+	}
 }
 
 // TODO(btracey): Think about making this an exported function when the
 // constraint interface is designed.
 func (f functionInfo) satisfies(method Method) error {
-	gradOk := f.IsGradient || f.IsFunctionGradient
-	if method.Needs().Gradient && !gradOk {
-		return errors.New("optimize: function does not implement needed Gradient or FunctionGradient")
+	if method.Needs().Gradient {
+		if method.Needs().Hessian {
+			// Method needs the gradient and the Hessian, any of these three interfaces will do.
```

Comment:
You are right. Some extra allocations will be necessary, but the information is there. This will simplify the checks, I will postpone the actual allocations and evaluation to the next PR.